packages/dev/core/src/Materials/Node/Blocks/debugBlock.ts

```diff
@@ -0,0 +1,113 @@
+import type { NodeMaterialBuildState } from "../nodeMaterialBuildState";
+import { RegisterClass } from "../../../Misc/typeStore";
+import { NodeMaterialBlock } from "../nodeMaterialBlock";
+import { NodeMaterialBlockConnectionPointTypes } from "../Enums/nodeMaterialBlockConnectionPointTypes";
+import type { NodeMaterialConnectionPoint } from "../nodeMaterialBlockConnectionPoint";
+import { NodeMaterialBlockTargets } from "../Enums/nodeMaterialBlockTargets";
+import { ShaderLanguage } from "core/Materials/shaderLanguage";
+import type { Scene } from "core/scene";
+
+/**
+ * Block used to render intermediate debug values
+ * Please note that the node needs to be active to be generated in the shader
+ * Only one DebugBlock should be active at a time
+ */
+export class NodeMaterialDebugBlock extends NodeMaterialBlock {
+    private _isActive = false;
+
+    /**
+     * Gets or sets a boolean indicating that the block is active
+     */
+    public get isActive(): boolean {
+        return this._isActive && this.debug.isConnected;
+    }
+
+    public set isActive(value: boolean) {
+        if (this._isActive === value) {
+            return;
+        }
+
+        this._isActive = value;
+    }
+
+    /**
+     * Creates a new NodeMaterialDebugBlock
+     * @param name defines the block name
+     */
+    public constructor(name: string) {
+        super(name, NodeMaterialBlockTargets.Fragment, true, true);
+
+        this.registerInput("debug", NodeMaterialBlockConnectionPointTypes.AutoDetect, true);
+
+        this.debug.excludedConnectionPointTypes.push(NodeMaterialBlockConnectionPointTypes.Matrix);
+    }
+
+    /** @internal */
+    public override get _isFinalOutputAndActive() {
+        return this.isActive;
+    }
+
+    /** @internal */
+    public override get _hasPrecedence() {
+        return true;
+    }
+
+    /**
+     * Gets the rgba input component
+     */
+    public get debug(): NodeMaterialConnectionPoint {
+        return this._inputs[0];
+    }
+
+    /**
+     * Gets the current class name
+     * @returns the class name
+     */
+    public override getClassName() {
+        return "NodeMaterialDebugBlock";
+    }
+
+    protected override _buildBlock(state: NodeMaterialBuildState) {
+        super._buildBlock(state);
+
+        if (!this._isActive) {
+            return this;
+        }
+
+        let outputString = "gl_FragColor";
+        if (state.shaderLanguage === ShaderLanguage.WGSL) {
+            outputString = "fragmentOutputs.color";
+        }
+
+        const debug = this.debug;
+        if (!debug.connectedPoint) {
+            return this;
+        }
+
+        if (debug.connectedPoint.type === NodeMaterialBlockConnectionPointTypes.Float) {
+            state.compilationString += `${outputString}  = vec4${state.fSuffix}(${debug.associatedVariableName}, ${debug.associatedVariableName}, ${debug.associatedVariableName}, 1.0);\n`;

```

Comment:
does the debug block needs a remap like feature to be able to look at floats ?

Reply:
It will be in the gray color so all good